Add reference sections and implement new childpages shortcode (#1293)
corda/corda-docs-portal: content/en/platform/corda/5.0/developing-applications/api/application/persistence.md

---
date: '2023-02-10'
title: "net.corda.v5.application.persistence"
menu:
  corda5:
    identifier: corda5-api-app-persistence
    parent: corda5-api-application
    weight: 6000
section_menu: corda5
---

The `persistence` package provides services for performing persistence operations; mainly reading and writing data to and from the database. The `PersistenceService` is the main service for providing this functionality.

Corda 5 supports CRUD (Create, Read, Update, Delete) operations for user-defined types. This is achieved using JPA-annotated entities and, to manage database migrations, Liquibase.

## Defining Custom Tables Using Liquibase Migrations

[CorDapps](../../../introduction/key-concepts.html#cordapps) store data in a relational database.

> When Corda creates a [virtual node](../../../introduction/key-concepts.html#virtual-nodes) for a CorDapp (as part of a [CPI](../../../introduction/key-concepts.html#corda-package-installer-cpi)), it requires associated tables, indexes, foreign-keys, etc.

Reviewer: ```suggestion
When Corda creates a [virtual node](../../../introduction/key-concepts.html#virtual-nodes) for a CorDapp (as part of a [CPI](../../../introduction/key-concepts.html#corda-package-installer-cpi)), it requires associated tables, indexes, foreign-keys, and so on.
```

To create these, you must embed Liquibase files into the CorDapp [CPK](../../../introduction/key-concepts.html#corda-packages-cpks).

> Liquibase manages DB changes in a “Change Log” which references one or more change sets.

Reviewer: ```suggestion
Liquibase manages database changes in a “Change Log” which references one or more change sets.
```

> You must specify the top level `databaseChangeLog` in a resource file in the CPK called `migration/db.changelog-master.xml`.

Reviewer: ```suggestion
You must specify the top-level `databaseChangeLog` in a resource file in the CPK called `migration/db.changelog-master.xml`.
```
